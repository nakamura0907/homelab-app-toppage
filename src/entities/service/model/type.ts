/**
 * サーバーのサービスモデル
 */
export type ServiceModel = {
  title: string;
  address: string;
};
