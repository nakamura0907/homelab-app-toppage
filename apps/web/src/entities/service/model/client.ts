import { ServiceModel } from "../types";

type ServiceClient = {
    fetchServices: () => Promise<ServiceModel[]>
}

const serviceListMock = [
    {
        title: 'Proxmox VE',
        address: 'https://192.168.0.200:8006',
    },
    {
        title: 'Pi-hole',
        address: 'http://192.168.0.231'
    },
    {
        title: 'Prometheus',
        address: 'http://192.168.0.232:9090'
    },
    {
        title: 'Loki',
        address: 'http://192.168.0.233:3100'
    }
] satisfies ServiceModel[];
const inMemoryClient: ServiceClient = {
    fetchServices: async () => {
        return serviceListMock
    }
}

/**
 * サービスクライアントを生成するファクトリ関数
 */
export const serviceClientFactory = (): ServiceClient => {
    return inMemoryClient
}
