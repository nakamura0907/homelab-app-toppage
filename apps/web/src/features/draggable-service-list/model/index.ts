export * from './use-draggable-service-list';
