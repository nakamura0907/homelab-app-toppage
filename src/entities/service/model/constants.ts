import { ServiceModel } from './type';

/**
 * サーバーのサービス一覧
 */
export const services = [
  {
    title: 'Proxmox',
    address: 'https://192.168.0.200:8006/',
  },
  {
    title: 'Pi-hole',
    address: 'http://192.168.0.231/',
  },
  {
    title: 'Grafana',
    address: 'http://192.168.0.241:3000/',
  },
  {
    title: 'Prometheus',
    address: 'http://192.168.0.241:9090/',
  },
] as const satisfies ServiceModel[];
