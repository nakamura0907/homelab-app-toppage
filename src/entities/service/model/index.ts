export type * from './type';

export * from './constants';
