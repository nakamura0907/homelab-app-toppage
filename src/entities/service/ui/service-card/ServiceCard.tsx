import { Card, CardHeader, CardTitle } from '@/components';
import { ServiceModel } from '../../model';
import Link from 'next/link';

type Props = {
  model: ServiceModel;
};

/**
 * サーバーのサービスカードコンポーネント
 */
export const ServiceCard: React.FC<Props> = ({ model }) => {
  return (
    <Card>
      <CardHeader>
        <CardTitle>
          <Link href={model.address} target="_blank" rel="noopener noreferrer">
            {model.title}
          </Link>
        </CardTitle>
      </CardHeader>
    </Card>
  );
};
