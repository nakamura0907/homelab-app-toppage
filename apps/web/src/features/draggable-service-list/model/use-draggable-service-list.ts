import { ServiceModel } from "@/entities/service";
import { arrayMove } from "@/shared";
import { useState } from "react";

/**
 * ドラッグ可能なサービス一覧のカスタムフック
 * 
 * @param initialServices 初期サービス一覧
 * @returns
 */
export const useDraggableServiceList = (initialServices: ServiceModel[]) => {
    const [services, setServices] = useState<ServiceModel[]>(initialServices);

    const changeServiceOrder = (oldIndex: number, newIndex: number) => {
        const newServices = arrayMove(services, oldIndex, newIndex);
        setServices(newServices);
    }

    return {
        services,
        /**
         * サービスの並び順を変更する
         * @param oldIndex 変更前のインデックス
         * @param newIndex 変更後のインデックス
         */
        changeServiceOrder,
    }
}