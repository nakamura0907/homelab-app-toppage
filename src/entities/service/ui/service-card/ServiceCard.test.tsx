import { cleanup, render, screen } from '@testing-library/react';
import { afterEach, describe, expect, it } from 'vitest';
import { ServiceCard } from './ServiceCard';
import { ServiceModel } from '../../model';

const mockServiceModel = {
  title: 'mock service',
  address: 'http://localhost:3000',
} satisfies ServiceModel;

describe('ServiceCard', () => {
  afterEach(() => {
    cleanup();
  });

  describe('render', () => {
    it('should match snapshot.', () => {
      const { container } = render(<ServiceCard model={mockServiceModel} />);

      expect(container).toMatchSnapshot();
    });
  });

  it('should render the title', () => {
    render(<ServiceCard model={mockServiceModel} />);

    expect(screen.getByText(mockServiceModel.title)).toBeInTheDocument();
  });

  it('should render a correct link', () => {
    render(<ServiceCard model={mockServiceModel} />);
    const link = screen.getByRole('link', { name: /mock service/i });

    expect(link).toHaveAttribute('href', mockServiceModel.address);
    expect(link).toHaveAttribute('target', '_blank');
    expect(link).toHaveAttribute('rel', 'noopener noreferrer');
  });
});
