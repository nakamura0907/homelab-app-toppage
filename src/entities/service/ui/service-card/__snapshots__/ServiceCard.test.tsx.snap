// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ServiceCard > render > should match snapshot. 1`] = `
<div>
  <div
    class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
    data-slot="card"
  >
    <div
      class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
      data-slot="card-header"
    >
      <div
        class="leading-none font-semibold"
        data-slot="card-title"
      >
        <a
          href="http://localhost:3000"
          rel="noopener noreferrer"
          target="_blank"
        >
          Nginx
        </a>
      </div>
    </div>
  </div>
</div>
`;
